### PR TITLE
feat: no-restricted-imports support casing

### PR DESCRIPTION
**Pull request by gfyoung**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What rule do you want to change?**

[no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports)

**Does this change cause the rule to produce more or fewer warnings?**

By default, the amount of warnings is unchanged.  Otherwise, it should be fewer warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

A new option is added, with a default value that maintains backwards-compatibility.

**Please provide some example code that this change will affect:**

```js
import 'food';
import 'fooBar';
```

```json
"no-restricted-imports": ["error", {
    "patterns": [{
      "group": ["foo[A-Z]*"]
    }]
}]
```

**What does the rule currently do for this code?**

It will issue warnings for both imports.

**What will the rule do after it's changed?**

If we add the new option:

```json
"no-restricted-imports": ["error", {
    "patterns": [{
      "group": ["foo[A-Z]*"],
      "caseSensitive": true
    }]
}]
```

The rule will issue a warning **ONLY** for the second import.

#### What changes did you make? (Give an overview)

Enhances [no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports) to support more granular import restrictions via case-sensitive pattern matching.  By default, this rule remains case-insensitive to maintain backwards compatibility with earlier versions.  Only by explicitly specifying it in a custom pattern object (see the added docs) can it be triggered.

Path matching in `.gitignore` can be case-sensitive, and `ignore@5.2.0` supports this as well as maintains support for relative paths, whose support was dropped after `4.0.6` but restored in `5.2.0` (this was a blocker for past upgrades).

#### Is there anything you'd like reviewers to focus on?

N/A

### CODE REVIEW

**Comment by mdjermanovic** [comment]
The option makes sense to me :+1:, and it would be nice to update the `ignore` dependency.

Waiting for more opinions from team members, since this is an enhancement.

File: lib/shared/ignore.js
Comment (mdjermanovic) [inline lib/shared/ignore.js:19]: I think it would be better to set `allowRelativePaths: true` directly in the rules where the `ignore` package is currently used. The current behavior of those two rules in regard to relative paths was not intended, so it seems better to avoid accidentally introducing the same in other places where we will use the `ignore` package in the future.

Comment (gfyoung) [reply lib/shared/ignore.js:19]: > The current behavior of those two rules in regard to relative paths was not intended

I'm not sure I understand.  Was supporting relative paths unintentional?

Comment (mdjermanovic) [reply lib/shared/ignore.js:19]: > Was supporting relative paths unintentional?

Yes, I believe it was unintentional in `no-restricted-imports` and `no-restricted-modules`, they were meant to work with bare specifiers. But, relative paths also worked because `ignore` didn't complain and users started using that functionality, so we can't remove it now. What I'm suggesting is to set `allowRelativePaths: true` directly in those two rules instead of enabling it by default in `shared/ignore.js`, because that mode isn't desirable.

Comment (gfyoung) [reply lib/shared/ignore.js]: Gotcha, done.

**Comment by gfyoung** [comment]
@mdjermanovic: PR has been updated with the requested changes, with tests still passing. 👍

**Comment by btmills** [comment]
This seems reasonable, but before adding new API, I find it helpful to understand the motivation. @gfyoung can you elaborate a bit on what use case you're trying to accomplish with case-sensitive path patterns?

**Comment by gfyoung** [comment]
@btmills: I believe the use case described in the PR description illustrates the motivation.

If not, please let me know what further information or detail is required.

**Comment by btmills** [comment]
@gfyoung the PR description covered a lot of the _how_ but not much of the _why_.  What are you hoping to do with ESLint that you can't today?

**Comment by gfyoung** [comment]
> What are you hoping to do with ESLint that you can't today?

I want to be able to do more granular import restrictions using case-sensitive regular expressions.

**Comment by btmills** [comment]
Okay, and what are you hoping to use case-sensitive import restrictions for?

**Comment by gfyoung** [comment]
The example provided in the [PR description](https://github.com/eslint/eslint/pull/15439#issue-806019528) illustrates a case where I would want to use case-sensitive import restrictions.

**Comment by nzakas** [comment]
@gfyoung the problem is that your description looks like a theoretical problem and not a real-life use case. What @btmills is asking for is something more concrete. What situation did you run into where this option would be helpful?

**Comment by gfyoung** [comment]
> the problem is that your description looks like a theoretical problem and not a real-life use case.

@nzakas: First off, expanding the API for the sake of a theoretical problem would be a waste of time for all of us.  I have been waiting to submit this PR for quite some time, as I had to get commits to `ignore` merged and released first.

Without sharing actual code, I thought the example using `foo` would be minimal and sufficient to explain the situation where this option would be helpful.  However, if you would like to see more realistic import names, here is an example, without sharing actual code that should help:

~~~js
import * as bots from bots;
import * as botUtils from botUtils;   // TODO: all imports should be from `bots`
~~~

In this case, we want all developers to importing from the `bots` file and not any of its constituent `bot[A-Z]*` files, which are in the process of being refactored and moved, but this will take quite some time to complete.

**Comment by nzakas** [comment]
@gfyoung please understand, we review a lot of PRs, and not all of them are well thought out. We need to ask followup questions and for more examples in order to figure out which ones are worth pursuing and which don’t provide much value to the project.

Answering our questions helps a lot, and if the PR description was enough, we wouldn’t ask for more info.

File: docs/rules/no-restricted-imports.md
Comment (nzakas) [inline docs/rules/no-restricted-imports.md:94]:
```suggestion
Pattern matches can also be configured to be case-sensitive:
```

File: lib/rules/no-restricted-imports.js
Comment (nzakas) [inline lib/rules/no-restricted-imports.js:68]: Can you set the default here for clarity?

**Review by nzakas — COMMENTED** [review]
Looks good, just a couple of suggestions.

**Comment by gfyoung** [comment]
@nzakas: I understand the need to request clarification.  As a maintainer myself, I can relate to the experience of working through PRs that have not been as well thought out as we would like.  That said, suggesting that someone's PR description is a "theoretical problem" before knowing more (e.g., not wanting to divulge actual code) is where some umbrage was taken.  IMO prefer to give contributors benefit of the doubt to prove themselves before coming to a conclusion about the PR.

Anyhow, I'll leave it at that.  Appreciate the feedback so far. 👍

File: lib/rules/no-restricted-imports.js
Comment (gfyoung) [reply lib/rules/no-restricted-imports.js:68]: I'm having trouble doing this.  When I tried:

```suggestion
                        type: "boolean",
                        default: false
```

I get a lot of errors like this during testing:

`Error: Schema for rule no-restricted-imports is invalid: default is ignored for: data3.caseSensitive`

File: docs/rules/no-restricted-imports.md
Comment (gfyoung) [reply docs/rules/no-restricted-imports.md]: Done

File: lib/rules/no-restricted-imports.js
Comment (mdjermanovic) [reply lib/rules/no-restricted-imports.js:68]: Ajv ignores defaults inside `anyOf` (https://github.com/eslint/eslint/pull/11427), which is the case here, thus this default wouldn't have any effects. RuleTester intentionally disallows ignored defaults (https://github.com/eslint/eslint/pull/11599) because relying on them can cause bugs, so we can't set default here.

Comment (gfyoung) [reply lib/rules/no-restricted-imports.js:68]: Ah, that explains it!

@nzakas: is there anything else you would want to do instead to clarify?

**Comment by mdjermanovic** [comment]
> Path matching in `.gitignore` can be case-sensitive, and `ignore@5.2.0` supports this as well as maintains support for relative paths, whose support was dropped after `4.0.6` but restored in `5.2.0` (this was a blocker for past upgrades)

Question: The `ignorecase` option we need for this PR was already available in `4.0.6`, but we must upgrade to `5.2.0` because of the caching bug that was fixed by https://github.com/kaelzhang/node-ignore/commit/f2f67cbaf9c8171bb2e63b1843f4ecdabc9ec6a1? (I noticed that the `invalid` test fails with `4.0.6`, but works when the `valid` test is removed).

**Comment by gfyoung** [comment]
Yes, that is in part why.

**Review by mdjermanovic — APPROVED** [review]
LGTM if accepted, thanks!

File: lib/rules/no-restricted-imports.js
Comment (nzakas) [reply lib/rules/no-restricted-imports.js:68]: Looks good.

**Comment by nzakas** [comment]
@gfyoung Im sorry that I didn’t explain well enough. We get a lot of people suggesting rules or changes to rules that are, in fact, theoretical or unlikely enough that it doesn’t make sense to accept. I did not mean to imply that your use case was theoretical, only that we needed more justification. Maintaining all of our rules, and ensuring things like rules don’t contradict each other, takes a lot work, so while we always assume positive intent from contributors (as we did here), we can’t assume that every case should be merged, which is why we asked for more details.

**Review by nzakas — APPROVED** [review]

**Comment by nzakas** [comment]
@btmills do you have the info you need here?

**Review by aladdin-add — APPROVED** [review]
LGTM, thanks! 💯

**Review by btmills — APPROVED** [review]
LGTM! Thanks for the additional explanation.

**Comment by gfyoung** [comment]
Fantastic! Thank you for all the feedback @btmills @mdjermanovic @nzakas 🙏

**Pull request merged**
